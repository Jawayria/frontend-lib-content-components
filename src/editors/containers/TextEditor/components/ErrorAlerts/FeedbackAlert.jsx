import React from 'react';
import PropTypes from 'prop-types';

import { Alert } from '@edx/paragon';
import { Info } from '@edx/paragon/icons';

export const hooks = {
  state: {
    isDismissed: (val) => React.useState(val),
  },
  dismissalHooks: ({ isError }) => {
    const [isDismissed, setIsDismissed] = hooks.state.isDismissed(false);
    React.useEffect(() => {
      setIsDismissed(isDismissed && !isError);
    },
    [isError]);
    return {
      isDismissed,
      dismissAlert: () => setIsDismissed(true),
    };
  },
};

export const FeedbackAlert = ({
  dismissError,
  isError,
  children,
}) => {
  const { isDismissed, dismissAlert } = hooks.dismissalHooks({ isError });
  if (!isError || isDismissed) {
    return null;
  }
  return (
    <Alert
      variant="danger"
      icon={Info}
      dismissible
      onClose={() => {
        dismissAlert();
        dismissError();
      }}
    >
      {children}
    </Alert>
  );
};

FeedbackAlert.propTypes = {
  isError: PropTypes.bool.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default FeedbackAlert;
