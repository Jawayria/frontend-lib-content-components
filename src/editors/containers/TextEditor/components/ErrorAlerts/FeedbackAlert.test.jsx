import React from 'react';
import { shallow } from 'enzyme';
import * as module from './FeedbackAlert';
import { MockUseState } from '../../../../../testUtils';

const { FeedbackAlert } = module;

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useRef: jest.fn(val => ({ current: val })),
  useEffect: jest.fn(),
  useCallback: (cb, prereqs) => ({ cb, prereqs }),
}));

const state = new MockUseState(module.hooks);
let hook;
const testValue = 'testVALUE';

describe('FeedbackAlert component', () => {
  describe('Hooks', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });
    describe('state hooks', () => {
      state.testGetter(state.keys.isDismissed);
    });
    describe('using state', () => {
      beforeEach(() => { state.mock(); });
      afterEach(() => { state.restore(); });
      describe('dismissalHooks', () => {
        beforeEach(() => {
          hook = module.hooks.dismissalHooks({ isError: testValue });
        });
        it('returns isDismissed value, initialized to false', () => {
          expect(state.stateVals.isDismissed).toEqual(hook.isDismissed);
        });
        test('dismissAlert sets isDismissed to true', () => {
          hook.dismissAlert();
          expect(state.setState.isDismissed).toHaveBeenCalledWith(true);
        });
        test('On Render, calls setIsDismissed', () => {
          expect(React.useEffect.mock.calls.length).toEqual(1);
          const [cb, prereqs] = React.useEffect.mock.calls[0];
          expect(prereqs[0]).toEqual(testValue);
          cb();
          expect(state.setState.isDismissed).toHaveBeenCalledWith(state.stateVals.isDismissed && !testValue);
        });
      });
    });
  });
  describe('Component', () => {
    describe('Snapshots', () => {
      let props;
      beforeAll(() => {
        props = {
          isError: false,
        };
        jest.spyOn(module.hooks, 'dismissalHooks').mockImplementation((value) => ({ isError: value }));
      });
      afterAll(() => {
        jest.clearAllMocks();
      });
      test('snapshot:  is Null when no error (FeedbackAlert)', () => {
        expect(shallow(<FeedbackAlert {...props}> <p> An Error Message </p></FeedbackAlert>)).toMatchSnapshot();
      });
      test('snapshot: Loads children and component when error (FeedbackAlert)', () => {
        expect(shallow(<FeedbackAlert {...props} isError> <p> An Error Message </p> </FeedbackAlert>)).toMatchSnapshot();
      });
    });
  });
});
