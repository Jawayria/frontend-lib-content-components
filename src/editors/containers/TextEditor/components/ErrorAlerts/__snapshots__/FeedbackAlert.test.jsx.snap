// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackAlert component Component Snapshots snapshot:  is Null when no error (FeedbackAlert) 1`] = `""`;

exports[`FeedbackAlert component Component Snapshots snapshot: Loads children and component when error (FeedbackAlert) 1`] = `
<Alert
  dismissible={true}
  onClose={[Function]}
  variant="danger"
>
   
  <p>
     An Error Message 
  </p>
   
</Alert>
`;
